pkg/tracing/google_cloud/google_cloud.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package google_cloud
+
+import (
+	"context"
+	"os"
+
+	cloudtrace "github.com/GoogleCloudPlatform/opentelemetry-operations-go/exporter/trace"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/version"
+	"github.com/thanos-io/thanos/pkg/tracing/migration"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/semconv"
+	"gopkg.in/yaml.v2"
+)
+
+// Config - YAML configuration.
+type Config struct {
+	ServiceName  string `yaml:"service_name"`
+	ProjectId    string `yaml:"project_id"`
+	SampleFactor uint64 `yaml:"sample_factor"`
+}
+
+// NewTracerProvider create tracer provider from YAML.
+func NewTracerProvider(ctx context.Context, logger log.Logger, conf []byte) (*tracesdk.TracerProvider, error) {
+	config := Config{}
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, err
+	}
+
+	exporter, err := cloudtrace.NewExporter(
+		cloudtrace.WithContext(ctx),
+		cloudtrace.WithProjectID(config.ProjectId),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return newTracerProvider(ctx, logger, tracesdk.NewBatchSpanProcessor(exporter),
+		config.SampleFactor, config.ServiceName), nil
+}
+
+func newTracerProvider(ctx context.Context, logger log.Logger, processor tracesdk.SpanProcessor, sampleFactor uint64, serviceName string) *tracesdk.TracerProvider {
+	fraction := 1 / float64(sampleFactor)
+	if sampleFactor == 0 {
+		fraction = 0
+	}
+
+	resource, err := resource.New(ctx, resource.WithAttributes(collectAttributes(serviceName)...))
+	if err != nil {
+		level.Warn(logger).Log("msg", "detecting resources for tracing provider failed", "err", err)
+	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithSpanProcessor(processor),
+		tracesdk.WithSampler(
+			migration.SamplerWithOverride(
+				tracesdk.ParentBased(tracesdk.TraceIDRatioBased(fraction)),
+				migration.ForceTracingAttributeKey,
+			),
+		),
+		tracesdk.WithResource(resource),
+	)
+
+	return tp
+}
+
+func collectAttributes(serviceName string) []attribute.KeyValue {
+	attr := []attribute.KeyValue{
+		semconv.ServiceNameKey.String(serviceName),
+		attribute.String("binary_revision", version.Revision),
+	}
+
+	if len(os.Args) > 1 {
+		attr = append(attr, attribute.String("binary_cmd", os.Args[1]))
```

Comment:
Why do we need to have binary Infos to attributes? Maybe remove this or at least make this configurable?

Reply:
Opentelemetry specification says lots of semantic_conventions , we should use it as possible. like `process.executable.name`. https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/process.md#process

Reply:
Honestly, this was purely to keep the behavior, attribute name etc. the same as it is now, as we wanted to make the change as 'backwards compatible' as possible. But yes, OTEL has more detailed guidance on semantic conventions and I'm definitely planning to follow as we move the rest of the instrumentation to OTEL. However, if the feeling is that this can be changed already here, I'm happy to follow through.